postinst: hardcode python binary and PYTHONPATH

orndorffgrant opened the pull request:
## Proposed Commit Message
<!-- Include a proposed commit message because all PRs can be merged in a variety of ways by the reviewer -->

```
postinst: hardcode python binary and PYTHONPATH

When the user's environment is set up to use a nonstandard python
installation, our postinst would fail. By specifying the exact python
binary to use, we avoid this problem. By also setting PYTHONPATH to the
location where our uaclient library is installed, we increase the
chances that we will succeed even if /usr/bin/python3 is overwritten.

If /usr/bin/python3 is overwritten with a different version of python,
whose standard library is incompatible with the expected version of
python, it is possible postinst could still fail.

Separately, if /usr/bin/python3 is overwritten with a different version
of python, the ua executable _will_ still fail.

LP: #1946034 #1944700 #1930121 #1938290 #1938074 #1927905
```

## Test Steps
<!-- Please include any steps necessary to verify (and reproduce if
this is a bug fix) this change on a live deployed system,
including any necessary configuration files, user-data,
setup, and teardown. Scripts used may be attached directly to this PR. -->
1. Run the new integration tests.
2. Forgive me for making Jenkins compile python from source for CI (This is the only "natural" way that I could find where user will end up with a nonstandard python install that causes the symptoms of the linked LP bugs).

## Desired commit type
<!-- put an `x` in one merge type to allow the reviewer to merge for you -->
 - [ ] Squash and merge: Using the "Proposed Commit Message"
 - [ ] Create a merge commit and use "Proposed Commit Message"
 - [x] Rebase and merge, no merge commit, rely only on existing commit messages

## Checklist
<!-- Go over all the following points, and put an `x` in all the boxes
that apply. -->
 - [ ] I have updated or added any unit tests accordingly
 - [x] I have updated or added any integration tests accordingly
 - [ ] I have updated or added any documentation accordingly

orndorffgrant commented on debian/ubuntu-advantage-tools.postinst:
```diff
@@ -16,6 +16,9 @@ fi
 . /usr/share/debconf/confmodule
 
+PYTHON=/usr/bin/python3
+export PYTHONPATH=/usr/lib/python3/dist-packages
```
This should only affect commands run as part of this script, but that is a good point that that is unnecessary, we could instead change this to
```
PYTHON=/usr/bin/env PYTHONPATH=/usr/lib/python3/dist-packages /usr/bin/python3
```

and then we wouldn't need the export line. Thoughts?

lucasmoura commented on debian/ubuntu-advantage-tools.postinst:
```diff
@@ -16,6 +16,8 @@ fi
 # https://lintian.debian.org/tags/postinst-does-not-load-confmodule.html
 . /usr/share/debconf/confmodule
 
+PYTHON="/usr/bin/env PYTHONPATH=/usr/lib/python3/dist-packages /usr/bin/python3"
```
Can we only have `/usr/lib/python3/`  as the python path here ?

orndorffgrant replied:
I just tested and unfortunately, having only `/usr/lib/python3/` doesn't work here

lucasmoura commented on features/util.py:
```diff
@@ -113,13 +114,24 @@ def repo_state_hash(include_features: bool = False):
     diff_out = subprocess.check_output(git_diff_cmd)
-    output_to_hash = rev_out + diff_out
+
+    git_status_cmd = ["git", "status", "--porcelain"]
```
I see why we are doing this, but I don't know if we should. Any test script that we create or any manual test we add, will trigger for a new package build. I know this is the more complete solution, but I don't know if it pays off.

What do you think @renanrodrigo @blackboxsw ?

orndorffgrant replied:
I updated this function to exclude more directories

renanrodrigo replied:
I think it is fine. We can tweak the "watched" directories in the future if needed, but so far, it works

lucasmoura commented on features/util.py:
```diff
@@ -88,38 +88,61 @@ def lxc_get_property(name: str, property_name: str, image: bool = False):
 
-def repo_state_hash(include_features: bool = False):
+def repo_state_hash(
+    exclude_dirs: List[str] = [".github", "demo", "features", "sru", "tools"]
```
Even though we are not doing this at the moment, but if we modify the `exclude_dirs` default list on the function, we will modify it on all other calls of `repo_state_hash`.

I know we are not doing that, but it is better to be defensive here. It is better to have that as `None`  by default and set it up with a default list.

lucasmoura commented on features/util.py:
```diff
@@ -88,38 +88,61 @@ def lxc_get_property(name: str, property_name: str, image: bool = False):
+        exclude = False
+        for exclude_dir in exclude_dirs:
+            if fname.startswith(exclude_dir):
+                exclude = True
```
Let's add a `break` here

orndorffgrant commented:
LP 1920121 is now the single bug for this issue.

I've updated the commit and fixed the mypy issue that stopped the last jenkins run. Let's let jenkins do a full run now.
